eduVPN/TwoFactorService.swift

```diff
@@ -0,0 +1,236 @@
+                let nsImage = NSImage(size: rep.size)
+                nsImage.addRepresentation(rep)
+                return nsImage
+            }
+        }
+        return nil
+    }
+    
+    /// Enroll user with TOTP 2FA at provider
+    ///
+    /// - Parameters:
+    ///   - info: Provider info
+    ///   - handler: Success or error
+    func enrollTotp(for info: ProviderInfo, secret: String, otp: String, handler: @escaping (Result<Void>) -> ()) {
+        let path: String = "two_factor_enroll_totp"
+        
+        guard let url = URL(string: path, relativeTo: info.apiBaseURL) else {
+            handler(.failure(Error.invalidProviderInfo))
+            return
+        }
+        
+        authenticationService.performAction(for: info) { (accessToken, idToken, error) in
+            guard let accessToken = accessToken else {
+                handler(.failure(error ?? Error.missingToken))
+                return
+            }
+            var request = URLRequest(url: url)
+            request.httpMethod = "POST"
+            request.httpBody = "totp_secret=\(secret)&totp_key=\(otp)".data(using: .utf8)
+            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
+            
+            let task = self.urlSession.dataTask(with: request) { (data, response, error) in
+                guard let data = data, let response = response as? HTTPURLResponse, 200..<300 ~= response.statusCode else {
+                    handler(.failure(error ?? Error.unknown))
+                    return
+                }
+                do {
+                    guard let json = try JSONSerialization.jsonObject(with: data, options: []) as? NSDictionary else {
+                        handler(.failure(Error.invalidTwoFactor))
+                        return
+                    }
+                    
+                    guard let instance = json.value(forKeyPath: path) as? [String: AnyObject] else {
+                        handler(.failure(Error.invalidTwoFactor))
+                        return
+                    }
+                    
+                    guard let isOk = instance["ok"] as? Bool else {
+                        handler(.failure(Error.invalidTwoFactor))
+                        return
+                    }
+                    
+                    if isOk {
+                        handler(.success(Void()))
+                    } else {
+                        guard let error = instance["error"] as? String else {
+                            handler(.failure(Error.invalidTwoFactor))
+                            return
+                        }
+                        
+                        switch error {
+                        case "user already enrolled":
+                            handler(.failure(Error.userAlreadyEnrolled))
+                        default:
+                            handler(.failure(Error.invalidTwoFactor))
+                        }
+                    }
```

Comment:
The `enrollYubico` and `enrollTotp` method seem to contain much of the same logic. Maybe consider moving those bits to a separate method?

Reply:
You are right. Moved bits to shared method.